File: javascript/dp-e07.js
const ALPHABET = {
    ".-" : "A", "-...": "B", "-.-.": "C", "-..": "D", ".": "E", "..-.": "F",
    "--.": "G", "....": "H", "..": "I", ".---": "J", "-.-": "K", ".-..": "L",
    "--": "M", "-.": "N", "---": "O", ".--.": "P", "--.-": "Q", ".-.": "R",
    "...": "S", "-": "T", "..-": "U", "...-": "V", ".--": "W", "-..-": "X",
    "-.--": "Y", "--..": "Z"
}

const morse2Letter = morse => ALPHABET[morse] ? ALPHABET[morse] : "?";
const morse2Word = morseWord => morseWord.split(' ').map(morse2Letter).join('');
const translate = morseSentence => morseSentence.map(morse2Word).join('');
const breakOnSlash = inp => inp.split(' / ');

function reverseObject(obj) {
  const keys = Object.keys(obj);
  return keys.reduce((result, key) => {
    result[obj[key]] = key;
    return result;
  }, {});
}

function decrypt(inp) {
  const morseSentence = breakOnSlash(inp);
  return translate(morseSentence);
}

function encrypt(inp) {
  const reversed = reverseObject(ALPHABET);
  const words = inp.toUpperCase().split(' ');
  const result = words.map((w) => w.split('').map((l) => reversed[l]).join(' '));
  return result.join(' / ');
}

if (process.argv[3] == '-d'){
  console.log(decrypt(process.argv[2]));
} else if (process.argv[3] == '-e') {
  console.log(encrypt(process.argv[2]));
} else {
  console.log("You need -d or -e.");
}

File: python/dp-e07.py
import sys

ALPHABET = {
    ".-" : "A", "-...": "B", "-.-.": "C", "-..": "D", ".": "E", "..-.": "F",
    "--.": "G", "....": "H", "..": "I", ".---": "J", "-.-": "K", ".-..": "L",
    "--": "M", "-.": "N", "---": "O", ".--.": "P", "--.-": "Q", ".-.": "R",
    "...": "S", "-": "T", "..-": "U", "...-": "V", ".--": "W", "-..-": "X",
    "-.--": "Y", "--..": "Z"
}

def morse2letter(morse):
    print(morse)
    try:
        return ALPHABET[morse]
    except KeyError as e:
        return "?"

def morse2word(morse_word):
    print(morse_word.split(' '))
    return ''.join([morse2letter(l) for l in morse_word.split(' ')])

def translate(morse_sentence):
    return ''.join([morse2word(w) for w in morse_sentence])

def break_on_slash(inp):
    return inp.split(' / ')

def decrypt(inp):
    morse_sentence = break_on_slash(inp)
    translated = translate(morse_sentence)
    return translated

def encrypt(inp):
    reverse = {v: k for k, v in ALPHABET.iteritems()}
    words = inp.upper().split(' ')
    res = []
    for w in words:
        res.append(' '.join([reverse[l] for l in list(w)]))

    return ' / '.join(res)





if __name__ == '__main__':
    try:
        if sys.argv[2] == '-d':
            print(decrypt(sys.argv[1]))
        elif sys.argv[2] == '-e':
            print(encrypt(sys.argv[1]))
    except IndexError as e:
        print("You are missing an argument.")
